comp/otelcol/otlp/config.go

```diff
@@ -62,17 +64,63 @@ func readConfigSection(cfg config.Reader, section string) *confmap.Conf {
 	for _, key := range cfg.AllKeysLowercased() {
 		if strings.HasPrefix(key, prefix) && cfg.IsSet(key) {
 			mapKey := strings.ReplaceAll(key[len(prefix):], ".", confmap.KeyDelimiter)
+
+			if v, ok := typeMap[mapKey]; ok {
+				if v == "bool" {
+					stringMap[mapKey] = cfg.GetBool(key)
+					continue
+				}
+				if v == "int" {
+					stringMap[mapKey] = cfg.GetInt(key)
+					continue
+				}
+			}
+
 			// deep copy since `cfg.Get` returns a reference
 			stringMap[mapKey] = deepcopy.Copy(cfg.Get(key))
 		}
 	}
 	return confmap.NewFromStringMap(stringMap)
 }
 
+// typeTagMap takes a struct and returns a map with mapstructure tag as key and field type as value.
+func typeTagMap(input interface{}, lead string, result map[string]string) map[string]string {
```

Comment:
This approach both feels very brittle, but also unrelated to the issue we were seeing earlier this week. Brittle because there's a lot the caller needs to do right (pass in a correctly constructed typeMap) and also because this isn't handling a lot of the general cases that typical reflection code needs to look at. Unrelated because this is handling implicit type coercion (which is itself brittle) but I recall the original issue was that the confMap input was being built with duplicate keys.